cmd/sfncli/runner.go

```diff
@@ -20,99 +25,170 @@ const maxTaskOutputLength = 32768
 	}
 
 	var taskInput map[string]interface{}
 	if err := json.Unmarshal([]byte(input), &taskInput); err != nil {
-		return t.handleProcessError(
-			ctx, "process-input", "unknown", fmt.Errorf("Input must be a json object: %s", err),
-		)
+		return t.sendTaskFailure(TaskFailureTaskInputNotJSON{input: input})
+	}
+
+	// convention: if the input contains _EXECUTION_NAME, pass it to the environment of the command
+	var executionName *string
+	if e, ok := taskInput["_EXECUTION_NAME"].(string); ok {
+		executionName = &e
+		t.logger.AddContext("execution_name", *executionName)
 	}
-	executionName, _ := taskInput["_EXECUTION_NAME"].(string)
 
 	marshaledInput, err := json.Marshal(taskInput)
 	if err != nil {
-		return t.handleProcessError(ctx, "process-input", executionName, err)
+		return t.sendTaskFailure(TaskFailureUnknown{fmt.Errorf("JSON input re-marshalling failed. This should never happen. %s", err)})
 	}
 
 	args = append(args, string(marshaledInput))
 
-	cmd := exec.CommandContext(ctx, t.cmd, args...)
-	cmd.Env = append(os.Environ(), "_EXECUTION_NAME="+executionName)
+	t.execCmd = exec.CommandContext(ctx, t.cmd, args...)
+	if executionName != nil {
+		t.execCmd.Env = append(os.Environ(), "_EXECUTION_NAME="+*executionName)
+	}
 
 	// Write the stdout and stderr of the process to both this process' stdout and stderr
 	// and also write to a byte buffer so that we can send the result to step functions
 	stderrbuf, _ := circbuf.NewBuffer(maxTaskFailureCauseLength)
 	stdoutbuf, _ := circbuf.NewBuffer(maxTaskOutputLength)
-	cmd.Stderr = io.MultiWriter(os.Stderr, stderrbuf)
-	cmd.Stdout = io.MultiWriter(os.Stdout, stdoutbuf)
+	t.execCmd.Stderr = io.MultiWriter(os.Stderr, stderrbuf)
+	t.execCmd.Stdout = io.MultiWriter(os.Stdout, stdoutbuf)
+
+	// forward signals to the command, handle SIGTERM
+	go t.handleSignals(ctx)
+
+	t.logger.InfoD("exec-command-start", logger.M{"args": args, "cmd": t.cmd})
+	if err := t.execCmd.Run(); err != nil {
+		stderr := strings.TrimSpace(stderrbuf.String()) // remove trailing newline
+		customErrorName := parseCustomErrorNameFromStdout(stdoutbuf.String())
+		if t.receivedSigterm {
+			if customErrorName != "" {
+				return t.sendTaskFailure(TaskFailureCustomErrorName{errorName: customErrorName, stderr: stderr})
+			}
+			return t.sendTaskFailure(TaskFailureCommandTerminated{stderr: stderr})
+		}
+		switch err := err.(type) {
+		case *os.PathError:
+			return t.sendTaskFailure(TaskFailureCommandNotFound{path: err.Path})
+		case *exec.ExitError:
+			status := err.ProcessState.Sys().(syscall.WaitStatus)
+			switch {
+			case status.Exited() && status.ExitStatus() > 0:
+				if customErrorName != "" {
+					return t.sendTaskFailure(TaskFailureCustomErrorName{errorName: customErrorName, stderr: stderr})
+				}
+				return t.sendTaskFailure(TaskFailureCommandExitedNonzero{stderr: stderr})
+			case status.Signaled() && status.Signal() == syscall.SIGKILL:
+				return t.sendTaskFailure(TaskFailureCommandKilled{stderr: stderr})
```

Comment:
Is the `customErrorName` intentionally ignored when the process is killed?

Reply:
yes... there's no way for a process to catch/handle SIGKILL, so a custom error message in stdout is unlikely. Also SIGKILL is usually triggered by the OS in exceptional circumstances, so it is most likely more significant than any custom error message the process might have managed to produce prior to getting SIGKILL

Reply:
I was imagining the case where an app does respond to SIGTERM but doesn't automatically exit but instead relies on the delayed SIGKILL.

Reply:
yeah that's a good call 4c0f232